chain/jsonrpc-primitives/src/types/config.rs

```diff
@@ -47,3 +53,503 @@ impl From<RpcProtocolConfigError> for crate::errors::RpcError {
+
+    // #############
+    // # Alt BN128 #
+    // #############
+    /// Base cost for multiexp
+    pub alt_bn128_g1_multiexp_base: super::Gas,
+    /// Per element cost for multiexp
+    pub alt_bn128_g1_multiexp_element: super::Gas,
+    /// Base cost for sum
+    pub alt_bn128_g1_sum_base: super::Gas,
+    /// Per element cost for sum
+    pub alt_bn128_g1_sum_element: super::Gas,
+    /// Base cost for pairing check
+    pub alt_bn128_pairing_check_base: super::Gas,
+    /// Per element cost for pairing check
+    pub alt_bn128_pairing_check_element: super::Gas,
+}
+
+/// This struct here exists to enforce that the value in the config is either
+/// `0` or `1`. We could have used a `bool` instead, but there's a chance that
+/// our current impl isn't perfect either and would need further tweaks in the
+/// future.
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
+pub enum StackLimiterVersion {
+    /// Old, buggy version, don't use it unless specifically to support old protocol version.
+    V0,
+    /// What we use in today's protocol.
+    V1,
+}
+
+impl StackLimiterVersion {
+    fn v0() -> StackLimiterVersion {
+        StackLimiterVersion::V0
+    }
+    fn repr(self) -> u32 {
+        match self {
+            StackLimiterVersion::V0 => 0,
+            StackLimiterVersion::V1 => 1,
+        }
+    }
+    fn from_repr(repr: u32) -> Option<StackLimiterVersion> {
+        let res = match repr {
+            0 => StackLimiterVersion::V0,
+            1 => StackLimiterVersion::V1,
+            _ => return None,
+        };
+        Some(res)
+    }
+}
+
+impl Serialize for StackLimiterVersion {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.repr().serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for StackLimiterVersion {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        u32::deserialize(deserializer).and_then(|repr| {
+            StackLimiterVersion::from_repr(repr)
+                .ok_or_else(|| serde::de::Error::custom("invalid stack_limiter_version"))
+        })
+    }
+}
+
+/// Describes limits for VM and Runtime.
+/// TODO #4139: consider switching to strongly-typed wrappers instead of raw quantities
+#[derive(Debug, Serialize, Deserialize, Clone, Hash, PartialEq, Eq)]
+pub struct VMLimitConfig {
+    /// Max amount of gas that can be used, excluding gas attached to promises.
+    pub max_gas_burnt: super::Gas,
+
+    /// How tall the stack is allowed to grow?
+    ///
+    /// See <https://wiki.parity.io/WebAssembly-StackHeight> to find out
+    /// how the stack frame cost is calculated.
+    pub max_stack_height: u32,
+    /// Whether a legacy version of stack limiting should be used, see
+    /// [`StackLimiterVersion`].
+    #[serde(default = "StackLimiterVersion::v0")]
+    pub stack_limiter_version: StackLimiterVersion,
```

Comment:
Presumably this can be just an `Option`? For all new fields we add in the future, I think they will be options in *this* struct, but not in the actual config. And if we are doing that for new fields, might as well do for fields which we added before?